### PR TITLE
[Bugfix] Start Assessment from Exercise Scores View

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Added back functionality to make an assessment for a specific submission.

### Description

- Added new Endpoint to lock a submission for a specific participation
- Added check in exercise scores view to be consistent with other exercise types
- Added check in exercise scores to not display View Assessment button when programming exercise allows no manual result
- New Endpoint is used in resolver for route

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration (you need a programming exercise with participations)
3. Go to Exercise Score view
4. Click on `View Assessment` button, you should see the assessment if it exists or get the lock for the assessment
5. Check how many locks you have -> Assessment Dashboard -> Assessments locked by you (see also Screenshots section)
image

### Screenshots

Assessments locked by me before trying to assess via exercise score view:
image

View Assessment button is now visible all the time and when clicking it and no assessment was yet started you will get the lock: 
image

After clicking on View Assessment button 2 times for not yet startet assessments:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 1]
```diff
@@ -636,6 +636,21 @@ public ProgrammingSubmission findByIdWithEagerResultAndFeedback(long submissionI
                 .orElseThrow(() -> new EntityNotFoundException("Programming submission with id \"" + submissionId + "\" does not exist"));
     }
 
+    /**
+     * Get the programming submission with the given ID from the database and lock the submission to prevent other tutors from receiving and assessing it.
+     *
+     * @param submissionId       the id of the programming submission
```
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:642]:
```suggestion
     * @param submissionId the id of the programming submission
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 2]
```diff
@@ -347,6 +348,43 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
         return ResponseEntity.ok().body(programmingSubmissions);
     }
 
+    // TODO: Make this call use submissionId instead of participationId (after implementation of one to many relation ship of submission and results)
+    /**
+     * GET /programming-submissions/:id : get the programmingSubmissions participation by it's id and locks the corresponding submission for assessment
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:353]:
```suggestion
     * GET /programming-submissions/:participationId/lock : get the programmingSubmissions participation by it's id and locks the corresponding submission for assessment
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 3]
```diff
@@ -347,6 +348,43 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
...
+        if (manualResult.isPresent()) {
+            return ResponseEntity.ok(participation);
+        }
+        else {
+            // Check lock limit
+            programmingSubmissionService.checkSubmissionLockLimit(exercise.getCourseViaExerciseGroupOrCourseMember().getId());
+
+            // As no manual result is present we need to lock the submission for assessment
+            Result latestAutomaticResult = participation.getResults().stream().filter(result -> result.getAssessmentType() == AssessmentType.AUTOMATIC)
+                    .sorted(Comparator.comparing(Result::getId).reversed()).iterator().next();
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:381]:
```suggestion
            Result latestAutomaticResult = participation.findLatestResult();
```

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html [hunk 4]
```diff
@@ -235,15 +241,15 @@ <h2>
...
+                            *ngIf="exercise.course"
                             [routerLink]="['/course-management', exercise.course?.id, 'programming-exercises', exercise.id, 'code-editor', value.participation.id, 'assessment']"
                             class="btn btn-primary btn-sm mr-1"
                         >
                             <fa-icon [icon]="'folder-open'" [fixedWidth]="true"></fa-icon>
                             {{ 'artemisApp.programmingSubmission.view' | translate }}
                         </button>
                         <button
-                            *ngIf="exercise.exerciseGroup && isManualResult(value)"
+                            *ngIf="exercise.exerciseGroup"
```
Comment (MaisiKoleni) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:252]: We also need to disable that for purely automatic programming exercise.

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
See inline. Code looks good otherwise.

**Review by airelawaleria — COMMENTED** [review]
Should be disabled for automatic prog. exercises

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html (hunk 4 above)
Comment (fde312) [reply src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html]: Fixed in `70289fe`

**Review by airelawaleria — APPROVED** [review]
Issue resolved. Re-approve

**Review by MaisiKoleni — APPROVED** [review]
Code looks good to me now and it works fine 👍

**Review by JohannesWeiss — APPROVED** [review]
Tested in testing session and worked great.

**Review by krusche — APPROVED** [review]

**Review by kloessst — APPROVED** [review]
Works fine.

**Review by EneaGore — APPROVED** [review]
Funcionatlity all fine. Locks and Assesments show as described. Minor thing, doing the Assesment from the Scores menu allows one to completly skip the Grading Intructions and the Check message for it.

**Pull request merged**